crate_universe/extension.bzl

```diff
@@ -0,0 +1,224 @@
+"""Module extension for generating third-party crates for use in bazel."""
+
+load("@bazel_tools//tools/build_defs/repo:git.bzl", "new_git_repository")
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+load("//crate_universe:defs.bzl", _crate_universe_crate = "crate")
+load("//crate_universe/private:crates_vendor.bzl", "CRATES_VENDOR_ATTRS", "generate_config_file", "generate_splicing_manifest")
+load("//crate_universe/private:generate_utils.bzl", "render_config")
+load("//crate_universe/private/module_extensions:cargo_bazel_bootstrap.bzl", "get_cargo_bazel_runner")
+
+def _generate_repo_impl(repo_ctx):
+    for path, contents in repo_ctx.attr.contents.items():
+        repo_ctx.file(path, contents)
+
+_generate_repo = repository_rule(
+    implementation = _generate_repo_impl,
+    attrs = dict(
+        contents = attr.string_dict(mandatory = True),
+    ),
+)
+
+def _generate_annotations(module_ctx, annotation_files):
+    """Given a list of annotation files, merge them into a single json object.
+
+    For an example json file, see examples/bzlmod/hello_world/annotations.json.
+    """
+    annotations = {}
+
+    for file in annotation_files:
+        content = json.decode(module_ctx.read(file))
+        for name, annotations_for_crate in content.items():
+            for annotation in annotations_for_crate:
+                annotation = _crate_universe_crate.annotation(**annotation)
+                annotations.setdefault(name, [])
+                annotations[name].append(annotation)
+
+    return annotations
+
+def _generate_hub_and_spokes(module_ctx, cargo_bazel, mod_path, cfg):
+    annotations = _generate_annotations(module_ctx, cfg.annotation_files)
```

Comment:
Do you think it makes more sense to do this processing of annotations in starlark here, or to just pass the JSON files to cargo-bazel and let it deal with them?

It feels like the starlark approach basically gives us stardoc at the expense of needing to duplicate the definitions across two languages and keep them in sync? And starlark gives us potentially earlier errors, but also potentially less control over the exact contents of the errors?

Reply:
Removing annotations from this PR, will do them in a follow-up CL to minimize size of this PR